userguide/payment/includes/payment-methods-flows.adoc

```diff
@@ -0,0 +1,93 @@
+
+image:https://github.com/killbill/killbill-docs/raw/v3/userguide/assets/img/payment-userguide/clientside-tokenization-payment-method-flow.png[align=center]
+
+. The end user enters payment details on the merchant website.
+
+. The merchant website converts it into a token and invokes the `addPaymentMethod` in Kill Bill with this token. This saves the token in the Kill Bill database.
+
+. Some payment gateways require the token to be saved within the gateway so as to be able to process recurring payments. In such cases, Kill Bill invokes the `addPaymentMethod` in the payment plugin corresponding to the payment gateway. 
+
+. The payment plugin then makes a gateway specific call to save the payment token within the gateway.
+
+=== Hosted Payment Page Payment Method Flow
+
+Sometimes, a merchant site may not provide a payment form/may not want to collect payment data and tokenize it. The hosted payment page flow is useful in such situations. In such cases, the merchant site needs to perform a two-step process. In the first step, it needs to invoke Kill Bill to obtain the URL to redirect the user and redirect the user to this page. Once the end-user enters his/her payment details on this page, the merchant site needs to invoke Kill Bill to add the payment method. Kill Bill then retrieves the payment details from the payment gateway and saves them in the Kill Bill database. This information is then used by Kill Bill to charge the customer.
+
+
+The following diagram explains this payment method flow:
+
+image:https://github.com/killbill/killbill-docs/raw/v3/userguide/assets/img/payment-userguide/hpp-payment-method-flow.png[align=center]
+
+. The end user initiates the checkout process.
+
+. The merchant website invokes a checkout servlet in Kill Bill. 
+
+. Kill Bill invokes the https://github.com/killbill/killbill-plugin-api/blob/d9eca5af0e37541069b1c608f95e100dbe13b301/payment/src/main/java/org/killbill/billing/payment/plugin/api/PaymentPluginApi.java#L269[PaymentPluginApi#buildFormDescriptor] method in the payment plugin. 
+
+. The `PaymentPlugin#buildFormDescriptor` method makes a gateway specific call and obtains a redirect URL.
+
+. The merchant website redirects the end user to this URL.
+
+. The end user enters payment details on this page.
+
+. If successful, the payment gateway redirects the user to a success URL on the merchant site.
+
+. The merchant site then invokes the `KillBill#addPaymentMethod`
+
+. Kill Bill in turn invokes the `PaymentPlugin#addPaymentMethod`.
+
+. The `PaymentPlugin#addPaymentMethod` performs a gateway specific call and obtains a gateway specific payment method id. This is then saved in the Kill Bill database.
+
+. Some payment gateways require the payment method id to be saved within the gateway so as to be able to process recurring payments. In such cases, the `PaymentPlugin#addPaymentMethod` may also perform a gateway specific call to save the payment method in the gateway.
```

Comment:
I don't think this applies here, as the payment method id / token was already fetched from the gateway.